Optimize workflows

mwperina opened the pull request:
# Run linters according to files changed in the PR

It's not needed to run linters on files, which are not changed within
the PR (for example if python sources are not changed, then it doesn't
make sense to install python linter dependencies and run the python
linter).

So this change introduces following lint jobs:

* common - executed on each patch (SPDX header and codespell)
* clang - executed only when C sources or meson files are changed
* md - executed only when markdown files are changed
* python - executed only when python files are changed

Signed-off-by: Martin Perina <mperina@redhat.com>


# Extract unit tests into standalone workflow

Unit tests execution was a bit hidden inside lint workflow, to make it
more visible it would be better to extract it into standalone workflow.

Signed-off-by: Martin Perina <mperina@redhat.com>


# Merge e2e and rpmbuild into single workflow

In the previous patches lint jobs and unit test was moved from build
workflow. After this change build worfklow contains only rpm build and
execution of e2e workflow as a part of build workflow. This execution
was a bit confusing (especially in github UI when looking at workflow
execution results for a PR), so it would be better to join rpmbuild and
e2e tests into a single workflow.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Run linters according to files changed in the PR

It's not needed to run linters on files, which are not changed within
the PR (for example if python sources are not changed, then it doesn't
make sense to install python linter dependencies and run the python
linter).

So this change introduces following lint jobs:

* common - executed on each patch (SPDX header and codespell)
* clang - executed only when C sources or meson files are changed
* md - executed only when markdown files are changed
* python - executed only when python files are changed

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,168 @@
+        uses: actions/upload-artifact@v3
+        with:
+          name: unit-test-logs
+          path: ./builddir/meson-logs/testlog-valgrind.txt
```
The unit tests are kind of hidden in the clang lint step (as previously in the lint step). I think it is fine when we run the unit tests always if some .c/.h file changes, but I am wondering it would be better to move them to their own step?!

mwperina replied:
Extracted unit tests into standalone workflow

mwperina committed:
Extract unit tests into standalone workflow

Unit tests execution was a bit hidden inside lint workflow, to make it
more visible it would be better to extract it into standalone workflow.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Merge e2e and rpmbuild into single workflow

In the previous patches lint jobs and unit test was moved from build
workflow. After this change build worfklow contains only rpm build and
execution of e2e workflow as a part of build workflow. This execution
was a bit confusing (especially in github UI when looking at workflow
execution results for a PR), so it would be better to join rpmbuild and
e2e tests into a single workflow.

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi reviewed (approved):
Looks great! :)
